pkg/controller/controller.go

```diff
@@ -924,7 +964,19 @@ func (p *csiProvisioner) Provision(ctx context.Context, options controller.Provi
 }
 
 func (p *csiProvisioner) setCloneFinalizer(ctx context.Context, pvc *v1.PersistentVolumeClaim) error {
-	claim, err := p.claimLister.PersistentVolumeClaims(pvc.Namespace).Get(pvc.Spec.DataSource.Name)
+	var claim *v1.PersistentVolumeClaim
+	var err error
+
+	if pvc.Spec.DataSourceRef != nil {
```

Comment:
How much of this logic can be simplified across all the data sources if we had a common function that can normalize DataSource and DataSourceRef? Something like:
```
func dataSource(claim *v1.PersistentVolumeClaim) (ObjectReference, error)
```

I see that this pattern repeats many times for each of the data source types.

Reply:
Changed to check the featuregate in the dataSource function. Also, since it is confirmed at the beginning of the prepareProvision function, the confirmation of featuregate removed in the "IsGrant ", "getSnapshotSource ", and "getPVCSource " function.

Reply:
It looks much better now thanks!